pkg/controller/externalippool/validate.go

```diff
@@ -83,47 +92,161 @@ func (c *ExternalIPPoolController) ValidateExternalIPPool(review *admv1.Admissio
+			}
+			mask = net.CIDRMask(int(subnetInfo.PrefixLength), 128)
+		}
+		subnet = &net.IPNet{
+			IP:   gatewayIP.Mask(mask),
+			Mask: mask,
 		}
-		mask = net.CIDRMask(int(subnetInfo.PrefixLength), 128)
 	}
-	subnet := &net.IPNet{
-		IP:   gatewayIP.Mask(mask),
-		Mask: mask,
+
+	// parse existing ip cidr and start-end ranges
+	existingCIDRs := make([]*net.IPNet, 0)
+	existingStartEndIPRanges := make([][2]net.IP, 0)
+	existingStartEndBigIntRanges := make([][2]*big.Int, 0)
+	for _, externalIPPool := range existingExternalIPPools {
+		for _, ipRange := range externalIPPool.Spec.IPRanges {
+			if ipRange.CIDR != "" {
+				cidr, _ := parseIPRangeCIDR(ipRange.CIDR)
+				existingCIDRs = append(existingCIDRs, cidr)
+			} else {
+				start, end, _ := parseIPRangeStartEnd(ipRange.Start, ipRange.End)
+				existingStartEndIPRanges = append(existingStartEndIPRanges, [2]net.IP{start, end})
+				existingStartEndBigIntRanges = append(existingStartEndBigIntRanges, [2]*big.Int{netutils.BigForIP(start), netutils.BigForIP(end)})
+			}
+		}
 	}
+
+	// validate ip cidr ranges
 	for _, ipRange := range ipRanges {
 		if ipRange.CIDR != "" {
-			_, cidr, err := net.ParseCIDR(ipRange.CIDR)
-			if err != nil {
-				return err.Error(), false
+			// parse cidr
+			cidr, errMsg := parseIPRangeCIDR(ipRange.CIDR)
+			if errMsg != "" {
+				return errMsg, false
 			}
-			if !ip.IPNetContains(subnet, cidr) {
-				return fmt.Sprintf("cidr %s must be a strict subset of the subnet", ipRange.CIDR), false
+
+			// validate if cidr is subset of given subnet info
+			if subnet != nil && !utilip.IPNetContains(subnet, cidr) {
+				return fmt.Sprintf("cidr %s must be a strict subset of the subnet %s", ipRange.CIDR, subnet.String()), false
 			}
-		} else {
-			start := net.ParseIP(ipRange.Start)
-			end := net.ParseIP(ipRange.End)
-			if !subnet.Contains(start) || !subnet.Contains(end) {
-				return fmt.Sprintf("IP range %s-%s must be a strict subset of the subnet", ipRange.Start, ipRange.End), false
+
+			// validate if cidr overlaps with any existing externalIPPool cidr
+			for _, existingCIDR := range existingCIDRs {
+				if cidr.Contains(existingCIDR.IP) || existingCIDR.Contains(cidr.IP) {
+					return fmt.Sprintf("cidr %s overlaps with cidr %s", cidr.String(), existingCIDR.String()), false
+				}
+			}
+
+			// validate if cidr overlaps with any existing externalIPPool start-end
+			for i := 0; i < len(existingStartEndIPRanges); i++ {
+				if cidr.Contains(existingStartEndIPRanges[i][0]) || cidr.Contains(existingStartEndIPRanges[i][1]) {
+					return fmt.Sprintf("cidr %s overlaps with range %s-%s",
+						cidr.String(), existingStartEndIPRanges[i][0].String(), existingStartEndIPRanges[i][1].String()), false
+				}
+			}
+
+			// add cidr to existing list to validate if cidrs within the current ip-ranges don't overlap
+			existingCIDRs = append(existingCIDRs, cidr)
+		}
+	}
+
+	// validate ip start-end ranges
+	for _, ipRange := range ipRanges {
+		if ipRange.CIDR == "" {
```

Comment:
I think this can be merged to the above loop via `else`?

Reply:
I wanted to collect all the CIDRs of existing pools and current pool before validating overlap of start-end ranges (of the current pool) with CIDRs

considering this example of external pool:
```
ipRanges:
	ipRange {start: 10.10.10.10, end: 10.10.10.20}
	ipRange {cidr: 10.10.10.0/24}
```
In the current logic if merge the for loops together this will pass. 

We can do this in a single iteration if we unify both CIDR and start-end range as you suggested!